usr/src/uts/common/fs/zfs/spa_history.c

```diff
@@ -192,6 +195,50 @@ spa_history_zone(void)
 }
 
 /*
+ * Post a history sysevent.  The nvlist_t* passed into this function will be
+ * duplicated (by nvlist_add_nvlist) and should be freed by caller.
+ * This function will only run in the context of the kernel.
+ */
+static void
+spa_history_log_notify(spa_t *spa, nvlist_t *nvl)
+{
+#ifdef _KERNEL
+	sysevent_t		*ev;
+	sysevent_attr_list_t	*attr = NULL;
+	sysevent_value_t	value;
+	sysevent_id_t		eid;
+
+	ev = sysevent_alloc(EC_ZFS, ESC_ZFS_HISTORY_EVENT,
+	    SUNW_KERN_PUB "zfs", SE_SLEEP);
+
+	value.value_type = SE_DATA_TYPE_STRING;
+	value.value.sv_string = spa_name(spa);
+	if (sysevent_add_attr(&attr, ZFS_EV_POOL_NAME, &value, SE_SLEEP) != 0)
+		goto done;
+
+	value.value_type = SE_DATA_TYPE_UINT64;
+	value.value.sv_uint64 = spa_guid(spa);
+	if (sysevent_add_attr(&attr, ZFS_EV_POOL_GUID, &value, SE_SLEEP) != 0)
+		goto done;
+
+	value.value_type = SE_DATA_TYPE_NVLIST;
+	value.value.sv_nvlist = nvl;
+	if (sysevent_add_attr(&attr, ZFS_EV_HISTORY, &value, SE_SLEEP) != 0)
+		goto done;
+
+	if (sysevent_attach_attributes(ev, attr) != 0)
+		goto done;
+	attr = NULL;
+
+	(void) log_sysevent(ev, SE_SLEEP, &eid);
+
+done:
+	sysevent_free_attr(attr);
+	sysevent_free(ev);
+#endif
+}
+
+/*
```

Comment:
This duplicates `spa_event_notify()` so it would make more sense to change `spa_event_notify()` to take an additional argument, `nvl` and add the `nvl` the same  way `vd` information is added.

```void
spa_event_notify(spa_t *spa, vdev_t *vd, nvlist_t *nvl, const char *name)```